### PR TITLE
add docs for GroupedMetadata and BaseMetadata

### PR DESCRIPTION
**Pull request by adriangb**

### CODE REVIEW

File: README.md
Comment (Zac-HD) [inline README.md:152]: I think I disagree here - for example if Pydantic has some metadata inheriting from BaseMetadata, that doesn't help Hypothesis resolve it at all; we'd still need specific knowledge.

(on the other hand inheriting from GroupedMetadata *does* help, precisely because it tells us to iterate over it and pluck out each BaseMetadata)

IMO the reason to have a BaseMetadata class at all is so that Hypothesis can tell at runtime when it's been passed things from a newer version of annotated-types than it understands, and request that the user update *Hypothesis*.

Comment (adriangb) [reply README.md:156]: I actually agree with you! I think I just worded poorly. In my mind a `GroupedMetadata` is to an `int` how a `List[int]` is to an `int`: it is not an `int` in and of itself. In other words, `GroupedMetada` _is not metadata_. Hence in my mind my wording of "If you are creating your own metadata, you _should_ inherit from `annotated_types.BaseMetadata`" made sense because if you are creating a `Field`.

I am also not saying that inheriting from `BaseMetadata` automatically makes consumers like Hypothesis be able to process this metadata. Instead, if lets Hypothesis identify that this is metadata and decide what to do with it (if Hypothesis chooses to ignore unknown subclasses of `BaseMetadata` or raise an error if it encounters them is up to Hypothesis).

File: README.md
Comment (adriangb) [inline README.md:174]:
```suggestion
In that case you should inherit from `GroupedMetadata` so that the grouping itself is not considered metadata (it is a _collection_ of metadata).
```

File: README.md
Comment (Zac-HD) [reply README.md:156]: > "If you are creating your own metadata, you _should_ inherit from `annotated_types.BaseMetadata`"

I meant that nobody outside the `annotated-types` package itself should inherit from `BaseMetadata`, because doing so means tools (e.g. Hypothesis) lose the ability to distinguish between the user installing a tool older than the current annotated-types version, vs the user installing something that also provides metadata types.

To me "noticing when there's metadata that I can't handle but expect to be able to" is the *only* sensible use of `BaseMetadata`; otherwise I'd have our metadata classes inherit directly from `object.

Put another way, I'd like to error on unknown metadata *if and only if* they're supplied by the `annotated-types` pacakge.

Comment (adriangb) [reply README.md:156]: > Put another way, I'd like to error on unknown metadata if and only if they're supplied by the annotated-types pacakge.

I see. My thought was that the tool (Hypothesis) in this case would error or ignore (up to the implementer) _any_ metadata it does not recognize, from `annotated-types` or not. If we want _different_ behavior for `annotated-types` vs. other user installed packages we'd need to add a level to the inheritance chain:

1. `BaseMetadata` namespaces anything in the `annotated-metadata` ecosystem from arbitrary stuff in `Annotated` (`Annotated[..., BaseMetadata(), "foo"]`).
2. `AnnotatedTypesMetadata` (or something like this) which namespaces all of the metadata provided by `annotated-types` itself
3. `Gt`, etc. are the concrete constraints.

```python
from annotated_types import Len, BaseMetadata, AnnotatedTypesMetadata

class PydanticPattern(BaseMetadata):
    ...

SomeField = Annotated[str, Len(1), PydanticPattern(...)]

# in some parser
args = list(get_args(SomeField))[1:]
for arg in args:
    if not isinstace(arg, BaseMetadata):
        continue
    metadata_type = type(arg)
    if metadata_type in KNOWN_METADATA_TYPES:
        ...  # do something
    # we don't know this metadata, up to the implementer to decide what to do
    # hypothesis could do the following
    if isinstance(arg, AnnotatedTypesMetadata):
        # unknown metadata coming from annotated-types
        raise ValueError("Please update hypothesis")
    # this is an unsupported package
    raise ValueError(f"Metadata of type {metadata_type} is not supported")
```

Comment (adriangb) [reply README.md:156]: Alternatively we could avoid the extra level of inheritance if Hypothesis is willing to check the module (`if type(arg).__module__.startswith("annotated_types")` or something like this).

Comment (Zac-HD) [reply README.md:156]: But why does Pydantic want to inherit from `BaseMetadata`?  The class doesn't actually do anything, and there's nothing that downstream libraries can do with that information if they don't have special knowledge of Pydantic anyway!

**Comment by Zac-HD** [comment]
OK, here's my current proposal:

- We remove *all* inheritance from the library
- We turn `Interval` into a *function* which returns `typing.Unpack[whatever, constraints]` (or a single constraint if there's only one)
  - actually... is `Annotated[T, *Foo]` or `Annotated[T, Unpack[Foo]]` valid?  If we need this changed we'd better be quick!
- We document that consumers of the annotated-types API are expected to iterate over `Unpack` members

I guess if `Unpack` doesn't work we can't do this.  I also don't have a good sense of how Pydantic 2 intends to represent `Field`s using annotated types - do you need something that can't be represented with `Annotated[T, <various things here?]`?

**Comment by adriangb** [comment]
> * We remove _all_ inheritance from the library
> * We turn `Interval` into a _function_ which returns `typing.Unpack[whatever, constraints]` (or a single constraint if there's only one)

Interesting, I was not aware of `typing.Unpack`, it's cool they added that! I think that would be a cool way to implement things in some cases (e.g. _our_ `Interval` could indeed become a function returning a `typing.Unpack`), but I think for something like Pydantic where `Field` is already a class (well it's a function returning a `FieldInfo` class) that also gets used as default values, I don't think that would be viable.

> I also don't have a good sense of how Pydantic 2 intends to represent `Field`s using annotated types

I would like @samuelcolvin to chime in, but my thought was that in Pydantic v2 `Field` would inherit from `annotated_types.GroupedMetadata` and unpack itself into standard metadata (aside from it's own `Regex` implementation and such).

>   * actually... is `Annotated[T, *Foo]` or `Annotated[T, Unpack[Foo]]` valid?  If we need this changed we'd better be quick!

I think both are valid

> * We document that consumers of the annotated-types API are expected to iterate over `Unpack` members

Agreed we should document this, if nothing else because we already use `Annotated[..., *Interval(...)]` and according to the 3.11 docs `typing.Unpack[...]` should be interpreted the same as if `*` had been used. 

All in all, here was my idea of how a consumer would generally use the inheritance we provide:

```python
from dataclasses import dataclass

from typing import Annotated, Any, Iterable, Iterator, get_args, get_origin, Unpack
from annotated_types import BaseMetadata, GroupedMetadata, Len


# a generic flattener, maybe we provide this in the future?
def flatten_metadata(args: Iterable[Any]) -> Iterable[BaseMetadata]:
    for arg in args:
        if isinstance(arg, BaseMetadata):
            yield arg
        elif isinstance(arg, GroupedMetadata):
            yield from arg
        elif get_origin(arg) is Unpack:
            packed_args = iter(get_args(arg))
            next(packed_args)
            yield from packed_args


# Pydantic classes
@dataclass
class Regex(BaseMetadata):
    pattern: str

@dataclass
class Alias(BaseMetadata):
    alias: str

@dataclass
class FieldInfo(GroupedMetadata):
    min_length: int | None = None
    regex: str | None = None
    alias: str | None = None

    def __iter__(self) -> Iterator[BaseMetadata]:
        if self.min_length is not None:
            yield Len(self.min_length)
        if self.regex is not None:
            yield Regex(self.regex)
        if self.alias is not None:
            yield Alias(self.alias)


def Field(
    min_length: int | None = None,
    regex: str | None = None,
    alias: str | None = None,
) -> Any:
    return FieldInfo(
        min_length=min_length,
        regex=regex,
        alias=alias,
    )


# user's code
# notice this is completely backwards compatible with Pydantic v1
PhoneNumber = Annotated[str, Field(min_length=10, alias="phoneNumber")]


# Pydantic's consumer
known_metadata = {
    Len,
    Regex,
    Alias,
}
def build_pydantic_core_schema_from_pep593_metadata(
    tp: type, allow_unknown: bool,
) -> Any:
    args = get_args(tp)
    assert args and args[0] is Annotated
    for metadata in flatten_metadata(args[1:]):
        if metadata in known_metadata:
            # build pydantic-core schema
            pass
        elif not allow_unknown:
            raise TypeError("unknown annotated-types metadata")
        else:
            pass


# Hypothesis' consumer
known_metadata = {
    Len,
    Regex,
    Alias,
}
supported_packages = {
    "annotated_types",
    "pydantic",
}
def _(
    tp: type
) -> Any:
    args = get_args(tp)
    assert len(args) > 2 and args[0] is Annotated
    for metadata in flatten_metadata(args[2:]):
        if metadata in known_metadata:
            # do thing
            continue
        if next(iter(metadata.__module__.split("."))) in supported_packages:
            raise RuntimeError(
                "Please update hypothesis"
            )
```

So `BaseMetadata` exists just to namespace metadata "in the annotated-types ecosystem" from arbitrary stuff, making it easier to write utility functions like `flatten_metadata` (that could exist here, in pydantic, in hypothesis or independently in each one). Yes, we could get rid of this namespace, but that makes it a bit harder to write reusable code because you can't layer your filtering/case handling and instead always have to check each object against an exhaustive list of all of the metadata/constraints you support.

While `GroupedMetadata` also servers the same use case to some extent, it also gives Pydantic a backwards compatible way to hook it's `Field` class into `annotated-types` in a way which requires no special handling from consumers. Even a consumer that errors on unknown metadata would still be able to handle a subset of `Field(...)` (as long as no Pydantic specific constraints are passed in). It also allows implementers the flexibility of having something that can be used both in PEP593 annotations and as default values (which is what Pydantic does and from discussions like [this one](https://github.com/jcrist/msgspec/issues/154) it may something others want to support as well). I don't think you can/would want to inherit from `typing.Unpack`, so that doesn't provide the same functionality.

**Comment by samuelcolvin** [comment]
Sorry I've been a bit AWOL on this. From a quick scan, looks good but let me know there's anything specific you need my help on.

**Comment by samuelcolvin** [comment]
Okay, I've now read your comment.

I'm happy for pydantic V2 `Field` to unpack itself to make it useful for other libraries.

The only edge case I can think of where this might not be desired is when users want to get the `Field` instance for secondary checks - if `Annotated[int, Field(..., gt=42, description='foobar')]` "unpacks itself", would it still be possible to retrieve the original `Field` instance?

**Comment by Zac-HD** [comment]
> ```python
> # a generic flattener, maybe we provide this in the future?
> def flatten_metadata(args: Iterable[Any]) -> Iterable[BaseMetadata]:
> ```

Remember that we also specify interpretations for `partial(operator.___, x)` and `slice` objects!

I'm kinda torn on the flattener; on one hand it's complicated enough that a flatten-and-canonicalize function would be pretty useful, on the other I want to keep the library as small as possible.  I think the argument that it's complicated wins here and we *should* ship it.

Note that this should actually return `Iterable[object]`: making the user responsible for filtering is a bit more work, but makes extending it to handle other types much easier.

> Yes, we could get rid of this [`BaseMetadata`] namespace, but that makes it a bit harder to write reusable code because you can't layer your filtering/case handling and instead always have to check each object against an exhaustive list of all of the metadata/constraints you support.

My argument is that you always need a mechanism to do the exhaustive check anyway, and should avoid having additional ways to check metadata (which could then get out of sync, etc).  Caching is a sufficient answer to perf concerns here IMO.

> While `GroupedMetadata` also servers the same use case to some extent (I am using it in an `isinstance` check above), it also gives Pydantic a backwards compatible way to hook it's `Field` class into `annotated-types` in a way which requires no special handling from consumers. 

If we specify that:

- Consumers of `annotated-types` metadata must iterate over `Unpack` items
- Pydantic fields must be written as `Annotated[T, *Field(...)]` or `Annotated[T, Unpack[Field(...)]]` (not `Annotated[T, Field(...)]`)
- Iterating over a `Field` instance yields annotated-types constraints (among whatever else seems useful)

then it will "just work" without requiring consumers to have any knowledge whatsoever of Pydantic.  However, the `Unpack[Field(...)]` syntax is _pretty darn awkward_, so I'd probably only support this if the `Annotated` type is generated by the library rather than hand-written by users 😕 

The best alternative is probably "just have and tell people to inherit from a `GroupedMetadata` class"; it's inelegant but it would work.

> Even a consumer that errors on unknown metadata would still be able to handle a subset of `Field(...)` (as long as no Pydantic specific constraints are passed in).

I don't really care about errors from such consumers, they're already living in a world where anyone can put anything in an `Annotated`.

> I'm happy for pydantic V2 `Field` to unpack itself to make it useful for other libraries.
> 
> The only edge case I can think of where this might not be desired is when users want to get the `Field` instance for secondary checks - if `Annotated[int, Field(..., gt=42, description='foobar')]` "unpacks itself", would it still be possible to retrieve the original `Field` instance?

You could have `Field.__iter__` yield from `(...constraints..., FieldWrapper[that_field])`?  Little awkward to avoid an infinite loop of unpacking 🤷‍♂️
 
I was imagining that the unpacked constraints would fully describe the Field though, including if necessary a "FieldResidue" object to hold all the other attributes.

**Comment by adriangb** [comment]
> Remember that we also specify interpretations for partial(operator.___, x) and slice objects!

Those are always going to be special cased, I was omitting for brevity.

> I'm kinda torn on the flattener; on one hand it's complicated enough that a flatten-and-canonicalize function would be pretty useful, on the other I want to keep the library as small as possible. I think the argument that it's complicated wins here and we should ship it.

I'm not arguing that we should ship it or not, I'm just sketching how it might look if we ship it or if Pydantic/Hypothesis implement it.

> My argument is that you always need a mechanism to do the exhaustive check anyway, and should avoid having additional ways to check metadata (which could then get out of sync, etc). Caching is a sufficient answer to perf concerns here IMO.

My concern isn't about performance but rather about usability. Yes, you always have to do some sort of exhaustive check at the end, but it's much easier to handle filtering, raising errors for unknown metadata, etc. when you can layer these things (e.g. filter out all metadata that is unrelated to `annotated-types`).

> Pydantic fields must be written as Annotated[T, *Field(...)] or Annotated[T, Unpack[Field(...)]] (not Annotated[T, Field(...)])

@samuelcolvin can comment in more detail but this would be breaking change for Pydantic

> You could have Field.__iter__ yield from (...constraints..., FieldWrapper[that_field])? Little awkward to avoid an infinite loop of unpacking 🤷‍♂️

> I was imagining that the unpacked constraints would fully describe the Field though, including if necessary a "FieldResidue" object to hold all the other attributes.

I'm lost here, sorry

**Comment by Zac-HD** [comment]
OK, yeah, never mind my speculative stuff, `GroupedMetadata` looks good to me.

**Comment by adriangb** [comment]
> The only edge case I can think of where this might not be desired is when users want to get the Field instance for secondary checks - if Annotated[int, Field(..., gt=42, description='foobar')] "unpacks itself", would it still be possible to retrieve the original Field instance?

I did the following test:

```python
from typing import Annotated, Any, get_args, get_type_hints
import typing

class ArbitraryIterable:
    def __iter__(self):
        return iter(["foo"])

def foo(
    unpack: Annotated[Any, typing.Unpack[ArbitraryIterable()]],
    star: Annotated[Any, *ArbitraryIterable()]
) -> Any:
    ...

print(get_args(get_type_hints(foo, include_extras=True)["unpack"])[1])  # *<__main__.ArbitraryIterable object at 0x102a94ad0>
print(get_args(get_type_hints(foo, include_extras=True)["star"])[1])  # foo
```

Maybe `Unpack` is just too new to have good support from `get_type_hints` and co, or maybe I'm doing something wrong.

In any case, I think it's a lot easier for us to say "consumers should iterate over any subclass of `GroupedMetadata` (which will work for any version of Python and requires nothing fancy at all) vs. saying "in Python 3.11+ use `*Field(...)` and in older versions use `Unpack[Field(...)]`, which is pretty funky, requires an import for users, is backwards incompatible for Pydantic and might break at any time (there are no guarantees as to the runtime behavior of `Unpack`).

**Review by Zac-HD — APPROVED** [review]
I don't think I'll personally check for `BaseMetadata`, but I've been convinced that this design makes sense.  Let's ship it!

And 💖 thanks again Adrian (and Samuel!) for the conversations, they can be wearing at times but I do think we get better designs out the other end.  Or in some cases _keep_ better designs 😅

**Comment by adriangb** [comment]
Discussions are good @Zac-HD, thank you for the pushback!

@samuelcolvin I'll let you review this before we merge since this will be the last PR before shipping this (after which it would be harder to turn back!)

**Review by samuelcolvin — COMMENTED** [review]
Okay, I've slept and consumed coffee and actually reviewed this. Sorry for hurried answer yesterday :see_no_evil:.

Over all I'm in favour of this except I don't think we should encourage libraries to inherit from `BaseMetadata` as argued by @Zac-HD. It's worth noting that we _could_ add that encouragement in a future release, but removing it/moving to discouraging inheritance from `BaseMetadata` is harder.

Still, @adriangb if you're convinced about this, I'm easy either way.

Encouraging inheritance from `GroupedMetadata` makes lots of sense - the other route would be to have some dunder attribute which identified a metadata group which wouldn't require `annotated-types` to be installed and imported in order to mark a metadata group. But since pydantic will have `annotated-types` as a required dependency, I'm happy with an abc as marker, it's definitely more canonical than a marker attribute.

File: README.md
Comment (adriangb) [reply README.md:156]: I think it would allow writing "generic" functions for the `annotated-types` "ecosystem" that do stuff like parse a type annotation into a list of metadata, which Hypothesis can then individually look at and consume, ignore, error, warn, or whatever it wants to do. Without that base class the only way to differentiate a piece of metadata is by exhaustive checking, which makes it harder to write generic utility functions because you have to pass around some enumeration of all known metadata classes as well as instructions as to what to do for stuff that isn't in that enumeration. I say this having written these sorts of parsers for `annotated-types` twice (once in our tests here and once in https://github.com/adriangb/typeval).

Comment (adriangb) [reply README.md:156]: And I think this generic functionality is important to enable building an ecosystem and not just a collection of metadata ( I realize we haven't really formally decided which one we are yet)

**Comment by adriangb** [comment]
> Over all I'm in favour of this except I don't think we should encourage libraries to inherit from `BaseMetadata` as argued by @Zac-HD. It's worth noting that we _could_ add that encouragement in a future release, but removing it/moving to discouraging inheritance from `BaseMetadata` is harder.
> 
> Still, @adriangb if you're convinced about this, I'm easy either way.

My opinion is that yes, we should encourage it, but since I'm in the minority here I'm happy to punt this down the road. My next question then is: should we explicitly tell others to _not_ inherit from the class and say that it should only be used to recognize metadata coming from `annotated-types`? Or should we just not document it at all (i.e. completely punt on it)?

**Comment by Zac-HD** [comment]
Let's punt on it, we can always work out what we're doing later 😆

File: README.md
Comment (Zac-HD) [reply README.md:156]: My understanding was that [we're right up the minimal-collection-of-metadata end *in this package*](https://github.com/annotated-types/annotated-types#design--history):

> ### Design & History
> This package was designed at the PyCon 2022 sprints by the maintainers of Pydantic and Hypothesis, with the goal of making it as easy as possible for end-users to provide more informative annotations for use by runtime libraries.
>
> **It is deliberately minimal**, and following PEP-593 allows considerable downstream discretion in what (if anything!) they choose to support. Nonetheless, we expect that staying simple and covering only the most common use-cases will give users and maintainers the best experience we can. If you'd like more constraints for your types - follow our lead, by defining them and documenting them downstream!

(emphasis added)

**Comment by adriangb** [comment]
Ok let's just remove that section. I'm taking a flight / vacation so not sure how soon I'll get around to it, if either of you want to push the changes please feel free!

**Pull request merged**
